okhttp/src/test/java/okhttp3/ServerTruncatesRequestTest.kt

```diff
@@ -0,0 +1,268 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3
+
+import okhttp3.Headers.Companion.headersOf
+import okhttp3.internal.duplex.AsyncRequestBody
+import okhttp3.internal.http2.ErrorCode
+import okhttp3.mockwebserver.MockResponse
+import okhttp3.mockwebserver.MockWebServer
+import okhttp3.mockwebserver.SocketPolicy
+import okhttp3.testing.PlatformRule
+import okhttp3.tls.internal.TlsUtil.localhost
+import okio.BufferedSink
+import okio.IOException
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.Assert.fail
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.rules.TestRule
+import org.junit.rules.Timeout
+import java.util.concurrent.TimeUnit
+
+class ServerTruncatesRequestTest {
+  @Rule @JvmField
+  val platform = PlatformRule()
+
+  @Rule @JvmField
+  val timeout: TestRule = Timeout(30_000, TimeUnit.MILLISECONDS)
+
+  @Rule @JvmField
+  val server = MockWebServer()
+
+  @Rule @JvmField
+  var clientTestRule = OkHttpClientTestRule()
+
+  private val listener = RecordingEventListener()
+  private val handshakeCertificates = localhost()
+
+  private var client = clientTestRule.newClientBuilder()
+      .eventListenerFactory(clientTestRule.wrap(listener))
+      .build()
+
+  @Before fun setUp() {
+    platform.assumeNotOpenJSSE()
+    platform.assumeHttp2Support()
+    platform.assumeNotBouncyCastle()
+  }
+
+  @Test fun serverTruncatesRequestOnLongPostHttp1() {
+    serverTruncatesRequestOnLongPost(https = false)
+  }
+
+  @Test fun serverTruncatesRequestOnLongPostHttp2() {
+    enableProtocol(Protocol.HTTP_2)
+    serverTruncatesRequestOnLongPost(https = true)
+  }
+
+  private fun serverTruncatesRequestOnLongPost(https: Boolean) {
+    server.enqueue(MockResponse()
+        .setSocketPolicy(SocketPolicy.DO_NOT_READ_REQUEST_BODY)
+        .setBody("abc")
+        .apply { this.http2ErrorCode = ErrorCode.NO_ERROR.httpCode })
+
+    val call = client.newCall(Request.Builder()
+        .url(server.url("/"))
+        .post(SlowRequestBody)
+        .build())
+
+    call.execute().use { response ->
+      assertThat(response.body!!.string()).isEqualTo("abc")
+    }
+
+    val expectedEvents = mutableListOf<String>()
+    expectedEvents += "CallStart"
+    expectedEvents += "ProxySelectStart"
+    expectedEvents += "ProxySelectEnd"
+    expectedEvents += "DnsStart"
+    expectedEvents += "DnsEnd"
+    expectedEvents += "ConnectStart"
+    if (https) {
+      expectedEvents += "SecureConnectStart"
+      expectedEvents += "SecureConnectEnd"
+    }
+    expectedEvents += "ConnectEnd"
+    expectedEvents += "ConnectionAcquired"
+    expectedEvents += "RequestHeadersStart"
+    expectedEvents += "RequestHeadersEnd"
+    expectedEvents += "RequestBodyStart"
+    expectedEvents += "RequestFailed"
```

Comment:
maybe a trailing comment here highlighting that the important part is we read the response headers and body even on error.